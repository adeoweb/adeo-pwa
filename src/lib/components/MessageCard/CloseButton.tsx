import React, { FunctionComponent } from 'react';

interface ICloseButtonProps {
    handleCloseMessage: () => void;
}

const CloseButton: FunctionComponent<ICloseButtonProps> = ({
    handleCloseMessage
}) => (
    <button
        type="button"
        className="close"
        aria-label="Close"
        onClick={handleCloseMessage}
    >
        <span aria-hidden="true">&times;</span>
    </button>
);

export default CloseButton;
