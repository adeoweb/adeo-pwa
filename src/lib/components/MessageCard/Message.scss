$padding-y: 0.7rem;
$padding-y-firs-last: $padding-y * 2;

.message {
    padding-top: $padding-y;
    padding-bottom: $padding-y;

    &:first-of-type {
        padding-top: $padding-y-firs-last;
    }

    &:last-of-type {
        padding-bottom: $padding-y-firs-last;
    }

    svg {
        height: 1rem;
    }
}
