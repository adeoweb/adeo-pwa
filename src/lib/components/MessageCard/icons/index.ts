export { default as errorIcon } from './error';
