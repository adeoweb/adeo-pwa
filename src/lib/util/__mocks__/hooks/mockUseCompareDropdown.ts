import { TUseCompareDropdown } from 'src/peregrine/lib/talons/adeoweb/Product/useCompareDropdown';

const compareDropdownTestRef = { current: null };

const mock: TUseCompareDropdown = {
    compareDropdownOpen: false,
    compareDropdownRef: compareDropdownTestRef
};

export default mock;
