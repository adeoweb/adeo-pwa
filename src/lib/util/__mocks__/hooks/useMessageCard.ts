import { IUseMessageCard } from 'src/peregrine/lib/talons/adeoweb/MessageCard/useMessageCard';

interface TMockUseMessageCard extends IUseMessageCard {
    addMessage: any;
    addMultipleMessages: any;
    clearBlock: any;
    clearAllMessages: any;
}

const mockMethods = {
    addMessage: () => {},
    addMultipleMessages: () => {},
    clearBlock: () => {},
    clearAllMessages: () => {}
};

export const mockWithMessages: TMockUseMessageCard = {
    messageBlocks: {
        error: ['test message', 'test message', 'test message'],
        warning: ['test message', 'test message'],
        success: ['test message'],
        info: []
    },
    ...mockMethods
};

export const mockNoMessages: TMockUseMessageCard = {
    messageBlocks: {
        error: [],
        warning: [],
        success: [],
        info: []
    },
    ...mockMethods
};
