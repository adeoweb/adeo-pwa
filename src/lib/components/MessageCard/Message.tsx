import React, { FunctionComponent } from 'react';
import { errorIcon } from 'src/lib/components/MessageCard/icons';
import MessageType from 'src/lib/constants/message';
import defaultClasses from './Message.scss';

interface IMessageProps {
    type: MessageType;
    message: string;
}

const getThemeIcon = (type: MessageType) => {
    switch (type) {
        case MessageType.ERROR:
            return errorIcon;
        default:
            return null;
    }
};

const Message: FunctionComponent<IMessageProps> = ({ type, message }) => (
    <p className={`mb-0 pl-4 pr-5 ${defaultClasses.message}`}>
        <strong>
            {getThemeIcon(type)}
            {message}
        </strong>
    </p>
);

export default Message;
