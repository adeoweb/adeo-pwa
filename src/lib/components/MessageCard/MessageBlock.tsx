import React, { FunctionComponent } from 'react';
import Message from './Message';
import CloseButton from './CloseButton';
import { useMessageCard } from 'src/peregrine/lib/talons/adeoweb/MessageCard/useMessageCard';
import MessageType from 'src/lib/constants/message';

interface IMessageBlock {
    type: MessageType;
    messages: string[];
}

const getThemeClass = (type: MessageType): string => {
    switch (type) {
        case MessageType.ERROR:
            return 'alert-danger';
        case MessageType.WARNING:
            return 'alert-danger';
        case MessageType.SUCCESS:
            return 'alert-success';
        case MessageType.INFO:
            return 'alert-info';
        default:
            return 'alert-info';
    }
};

const renderMessages = ({
    type,
    messages
}: {
    type: MessageType;
    messages: string[];
}) =>
    messages.map((message, index) => {
        if (!message) {
            return null;
        }

        return (
            <Message
                key={`${index}-${message}`}
                type={type}
                message={message}
            />
        );
    });

const MessageBlock: FunctionComponent<IMessageBlock> = ({ type, messages }) => {
    const { clearBlock } = useMessageCard();

    const handleClearBlock = () => {
        clearBlock({ type });
    };

    return (
        <div className={`position-relative ${getThemeClass(type)}`}>
            {renderMessages({ type, messages })}
            <CloseButton handleCloseMessage={handleClearBlock} />
        </div>
    );
};

export default MessageBlock;
