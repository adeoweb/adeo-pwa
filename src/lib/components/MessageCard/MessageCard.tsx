import React, { FunctionComponent } from 'react';

import MessageType, { MessageTypeKeys } from 'src/lib/constants/message';
import { IMessageBlocks } from 'src/peregrine/lib/store/reducers/adeoweb/messageCard';
import { useMessageCard } from 'src/peregrine/lib/talons/adeoweb/MessageCard/useMessageCard';

import MessageBlock from './MessageBlock';

const shouldShowCard = (messageBlocks: IMessageBlocks) =>
    Boolean(
        messageBlocks[MessageType.INFO].length ||
            messageBlocks[MessageType.ERROR].length ||
            messageBlocks[MessageType.SUCCESS].length ||
            messageBlocks[MessageType.WARNING].length
    );

const renderMessageBlocks = (messageBlocks: IMessageBlocks) => {
    const keys = Object.keys(MessageType) as MessageTypeKeys[];

    return keys.map((key, index) => {
        const messageKey = MessageType[key];

        if (!messageBlocks[messageKey].length) {
            return null;
        }

        return (
            <MessageBlock
                key={`${index}-${messageKey}`}
                type={messageKey}
                messages={messageBlocks[messageKey]}
            />
        );
    });
};

const MessageCard: FunctionComponent = () => {
    const { messageBlocks } = useMessageCard();
    const showCard = shouldShowCard(messageBlocks);
    const display = showCard ? 'd-block' : 'd-none';

    if (showCard) {
        window.scrollTo(0, 0);
    }

    return (
        <div
            className={`ss-card alert alert-dismissible fade show p-0 ${display}`}
            role="alert"
        >
            {renderMessageBlocks(messageBlocks)}
        </div>
    );
};

export default MessageCard;
