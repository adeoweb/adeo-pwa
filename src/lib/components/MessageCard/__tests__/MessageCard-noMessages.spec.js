import React from 'react';
import testRenderer from 'react-test-renderer';
import MessageCard from '../MessageCard';
import { mockNoMessages } from 'src/lib/util/__mocks__/hooks/useMessageCard';

jest.mock('src/peregrine/lib/talons/adeoweb/MessageCard/useMessageCard', () => {
    return {
        useMessageCard: () => mockNoMessages
    };
});

test('Matches snapshot', () => {
    const component = testRenderer.create(<MessageCard />);
    expect(component.toJSON()).toMatchSnapshot();
});
