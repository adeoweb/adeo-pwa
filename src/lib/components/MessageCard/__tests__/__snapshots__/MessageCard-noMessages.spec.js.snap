// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="ss-card alert alert-dismissible fade show p-0 d-none"
  role="alert"
/>
`;
